utils/updateable_value.hh

```diff
@@ -0,0 +1,212 @@
+
+namespace utils {
+
+// This file contains two templates, updateable_value_source<T> and updateable_value<T>.
+//
+// The two are analogous to T and const T& respectively, with the following additional
+// functionality:
+//
+//  - updateable_value contains a copy of T, so it can be accessed without indirection
+//  - updateable_value and updateable_value_source track each other, so if they move,
+//    the references are updated
+//  - an observe() function is provided (to both) that can be used to attach a callback
+//    that is called whenever the value changes
+
+template <typename T>
+class updateable_value_source;
+
+class updateable_value_source_base;
+
+// Base class for updateable_value<T>, containing functionality for tracking
+// the update source. Used to reduce template bloat and not meant to be used
+// directly.
+class updateable_value_base {
+protected:
+    const updateable_value_source_base* _source = nullptr;
+public:
+    updateable_value_base() = default;
+    explicit updateable_value_base(const updateable_value_source_base& source);
+    ~updateable_value_base();
+    updateable_value_base(const updateable_value_base&);
+    updateable_value_base& operator=(const updateable_value_base&);
+    updateable_value_base(updateable_value_base&&) noexcept;
+    updateable_value_base& operator=(updateable_value_base&&) noexcept;
+
+    friend class updateable_value_source_base;
+};
+
+
+// A T that can be updated at runtime; uses updateable_value_base to track
+// the source as the object is moved or copied. Copying across shards is supported.
+template <typename T>
+class updateable_value : public updateable_value_base {
+    T _value = {};
+private:
+    const updateable_value_source<T>* source() const;
+public:
+    updateable_value() = default;
+    explicit updateable_value(T value) : _value(std::move(value)) {}
+    explicit updateable_value(const updateable_value_source<T>& source);
+    updateable_value(const updateable_value& v);
+    updateable_value& operator=(const updateable_value&);
+    updateable_value(updateable_value&&) noexcept;
+    updateable_value& operator=(updateable_value&&) noexcept;
+    const T& operator()() const { return _value; }
+    operator const T& () const { return _value; }
+    const T& get() const { return _value; }
+    observer<T> observe(std::function<void (const T&)> callback) const;
+
+    friend class updateable_value_source_base;
+    template <typename U>
+    friend class updateable_value_source;
+};
+
+// Contains the mechanisms to track updateable_value_base.  Used to reduce template
+// bloat and not meant to be used directly.
+class updateable_value_source_base {
+protected:
+    // This class contains two different types of state: values and
+    // references to updateable_value_base. We consider adding and removing
+    // such references const operations since they don't change the logical
+    // state of the object (they don't allow changing the carried value).
```

Comment:
Any caller of those operations can observe `_refs` changing as a result of the call.  When a method changes observable functioning of an object, it's a rather curious decision to mark it `const`.  I find the division of state into "two different types" arbitrary and unwarranted.

Reply:
These objects are embedded into a db::config object, and into other intermediate configuration objects. Right now, we can pass db::config references (and those intermediate configuration objects) as const, which makes sense, because the consumer of configuration doesn't change it.

If we make reference tracking non-const, we'll have to pass configuration as non-const. When readers then ask why the consumer needs to mutate the configuration (rather than just read it), we'll need to explain that it doesn't mutate the configuration values, but instead so that internal reference tracking within the configuration can update its tracking state. Instead of having to explain it every time we pass configuration around, we explain it here once. And the justification is that there really are two different types of state, one is the payload, and one is tracking the association of references to the payload. One can add a reference to the payload without having the ability to change the payload.

Without mutable we'd have
 - can change payload and add reference (non-const)
 - can neither change payload nor add reference (const)

With mutable we have
 - can change payload and add reference (non-const)
 - cannot change payload, but still can add reference (const)

The former makes const unusable, the latter makes it useful.

Reply:
(in a similar way, copying a const shared_ptr causes an observable change (use_count() and unique() will change), yet it is allowed because the original object's primary state does not change; it's just bookkeeping state that changes)